File: LeetCodeSolutions/LeetCode_0051.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;
/*
 * @lc app=leetcode.cn id=51 lang=cpp
 *
 * [51] N皇后
 */

// @lc code=start
class Solution {
public:
    vector<vector<string> > ret;
    vector<vector<string> > solveNQueens(int n) {
        vector<vector<int> > counter(n, vector<int>(n, 0));
        // cout << string(n, '.') << endl;
        vector<string> tmp(n, string(n, '.'));

        helper(tmp, counter, 0, n);
        return ret;
    }

    void helper(vector<string> &tmp, vector<vector<int> > &cnt, int row, int n){
        // for(auto s: tmp){
        //     cout << s << endl;
        // }
        // cout << endl;
        if (row == n) {ret.push_back(tmp);return;}
        for(int i = 0; i < n; ++i){
            if (cnt[row][i] == 0){
                for (int j = 0; j < n; ++j) ++cnt[j][i];
                for (int l = 1; l < n; ++l){
                    int ii = row - l, jj = i - l;
                    if (ii >=n || jj >= n || ii < 0 || jj < 0) break;
                    ++cnt[ii][jj];
                }
                for (int l = 1; l < n; ++l){
                    int ii = row + l, jj = i + l;
                    if (ii >=n || jj >= n || ii < 0 || jj < 0) break;
                    ++cnt[ii][jj];
                }
                for (int l = 1; l < n; ++l){
                    int ii = row - l, jj = i + l;
                    if (ii >=n || jj >= n || ii < 0 || jj < 0) break;
                    ++cnt[ii][jj];
                }
                for (int l = 1; l < n; ++l){
                    int ii = row + l, jj = i - l;
                    if (ii >=n || jj >= n || ii < 0 || jj < 0) break;
                    ++cnt[ii][jj];
                }
                tmp[row][i] = 'Q';
                helper(tmp, cnt, row + 1, n);
                tmp[row][i] = '.';
                for (int j = 0; j < n; ++j) --cnt[j][i];
                for (int l = 1; l < n; ++l){
                    int ii = row - l, jj = i - l;
                    if (ii >=n || jj >= n || ii < 0 || jj < 0) break;
                    --cnt[ii][jj];
                }
                for (int l = 1; l < n; ++l){
                    int ii = row + l, jj = i + l;
                    if (ii >=n || jj >= n || ii < 0 || jj < 0) break;
                    --cnt[ii][jj];
                }
                for (int l = 1; l < n; ++l){
                    int ii = row - l, jj = i + l;
                    if (ii >=n || jj >= n || ii < 0 || jj < 0) break;
                    --cnt[ii][jj];
                }
                for (int l = 1; l < n; ++l){
                    int ii = row + l, jj = i - l;
                    if (ii >=n || jj >= n || ii < 0 || jj < 0) break;
                    --cnt[ii][jj];
                }
            }
        }
    }
};
// @lc code=end

File: LeetCodeSolutions/LeetCode_0051.py
#
# @lc app=leetcode.cn id=51 lang=python3
#
# [51] N皇后
#

# @lc code=start
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        visited = set()
        queen_pos = []

        def helper(idx, ret):
            if idx == n:
                queen_pos.append(ret[:])
                return
            for col in range(n):
                if (idx, col) not in visited:
                    cur_pos = [(idx, col)]
                    visited.add((idx, col))
                    for [dx, dy] in [[-1, 0], [1, 0], [0, -1], [0, 1], [-1, -1], [1, 1], [1, -1], [-1, 1]]:
                        cx, cy = idx + dx, col + dy
                        while -1 < cx < n and -1 < cy < n:
                            if (cx, cy) not in visited:
                                visited.add((cx, cy))
                                cur_pos.append((cx, cy))
                            cx += dx
                            cy += dy
                    ret.append(col)
                    helper(idx + 1, ret)
                    ret.pop()
                    while cur_pos:
                        visited.remove(cur_pos.pop())

        helper(0, [])
        # n_ret = len(queenpos)
        ret = []
        for cols in queen_pos:
            tmp = []
            for idx in cols:
                s = ''
                for i in range(n):
                    if i == idx:
                        s += 'Q'
                    else:
                        s += '.'
                tmp.append(s)
            ret.append(tmp)
        return ret

# @lc code=end
